corehq/apps/reports/tests/test_scheduled_reports.py

```diff
@@ -0,0 +1,153 @@
+from contextlib import contextmanager
+
+from django.test import TestCase
+from django.urls import reverse
+
+from corehq import privileges
+from corehq.apps.commtrack.tests.util import bootstrap_domain
+from corehq.apps.es.case_search import case_search_adapter
+from corehq.apps.es.client import manager
+from corehq.apps.es.forms import form_adapter
+from corehq.apps.es.tests.utils import es_test
+from corehq.apps.es.users import user_adapter
+from corehq.apps.locations.tests.util import setup_locations_and_types
+from corehq.apps.reports.views import ScheduledReportsView
+from corehq.apps.saved_reports.models import ReportConfig, ReportNotification
+from corehq.apps.users.dbaccessors import delete_all_users
+from corehq.apps.users.models import HqPermissions, WebUser
+from corehq.apps.users.models_role import UserRole
+from corehq.util.test_utils import disable_quickcache, privilege_enabled
+
+
+@es_test(requires=[
+    case_search_adapter,
+    form_adapter,
+    user_adapter,
+], setup_class=True)
+@disable_quickcache
+@privilege_enabled(privileges.API_ACCESS)
+class TestLocationRestrictedScheduledReportViews(TestCase):
+    domain = 'test-report-scheduling'
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.domain_obj = bootstrap_domain(cls.domain)
+        location_types, locations = setup_locations_and_types(
+            cls.domain,
+            location_types=['city'],
+            stock_tracking_types=[],
+            locations=[('Bhopal', []), ('Delhi', [])],
+        )
+
+        cls.user_location = locations['Bhopal']
+
+    @classmethod
+    def tearDownClass(cls):
+        super().tearDownClass()
+        cls.domain_obj.delete()
+        delete_all_users()
+
+    def test_location_restricted_user_can_access_reports_home_view(self):
+        with self._get_web_user_with_permissions_added():
+            url = reverse('reports_home', args=(self.domain,))
+            response = self.client.get(url)
+            self.assertEqual(response.status_code, 302)
+
+    def test_location_restricted_user_can_access_scheduled_report_view(self):
+        with self._get_web_user_with_permissions_added():
+            url = reverse(ScheduledReportsView.urlname, args=(self.domain,))
+            response = self.client.get(url)
+            self.assertEqual(response.status_code, 200)
```

Comment:
I am confused for why this is working without the feature flag enabled. I assume that now since we have added the check for feature flag to be enabled, the views intended should not be accessible for location restricted user unless the flag is enabled.

Reply:
Have made change to the tests to disable `access_all_locations` permission for the web user. Also, added feature flag to the existing tests. To verify the usage of FF, added a test where FF is not enabled and it failes with `403` as expected. Ref. https://github.com/dimagi/commcare-hq/pull/32964/files#diff-1bd0dcd0926f8d64a7a2fb1651f5a7fdf8df3f36b6b6f3fb3f6ec1fce6a798ebR51-R62

@mkangia Now there are 3 possible scenarios for each of the view, where request is made by:
1. User with `access_all_locations` permission enabled. Here FF status is not relevant.
2. User with  `access_all_locations` permission disabled and FF enabled.
3. User with  `access_all_locations` permission disabled and FF disabled.

@mkangia Should we add 3 such tests for each of the view, or can add them for one of the view only for confirmation?

Reply:
We should add for each of the view. You could also extract this into a single view and inherit each view with it to write only 1 set of test but that seems a bit risky.

I believe the tests we need are
User with access_all_locations permission disabled and FF enabled.
User with access_all_locations permission disabled and FF disabled.

Reply:
https://github.com/dimagi/commcare-hq/pull/32964/commits/5a5f0f9910d73832b916c9ce9c629350599f5685